src/ruby.rs

```diff
@@ -787,6 +787,36 @@ extern {
     ///
     //+ c-func: variable.c `const char *rb_obj_classname(VALUE)`
     pub fn rb_obj_classname(obj: VALUE) -> *const c_char;
+
+    /// Raises a Ruby exception of the provided class with message.
+    ///
+    /// * `exc` - a Ruby exception class (e.g. [`RuntimeError`](rb_eRuntimeError),
+    ///     [`TypeError`](rb_eTypeError))
+    /// * `message` - a Ruby format string, called in `printf` style
+    ///     * additional arguments are applied as in `printf`
+    ///
+    /// # Safety
+    ///
+    /// * Since `rb_raise` will stop execution of the current context, special care
```

Comment:
This function is highly unsafe. Internally, Ruby exceptions are implemented using `longjmp`(https://en.wikipedia.org/wiki/Setjmp.h). Jumping out of a Rust frame is considered undefined behavior, as it bypasses Rust's memory management and stack unwinding behavior. This could result in crashes or memory corruption. There are currently no officially sanctioned way to invoke a C function that does a `longjmp` such as `rb_raise`.

At minimum, this should only be called at the very edge of the FFI boundary after explicitly dropping any droppable references:

```rust
extern "c" fn some_ruby_method(rb_self: VALUE) -> VALUE {
    let result = {
        let droppable = SomeStruct::new();
        let is_success = droppable.do_work();

        if is_success {
            Ok(Qtrue);
        } else {
            // note that this only really works if the message is a static C-string!
            // If the message came from a heap-allocated `String`, for example,
            // that itself will need to be dropped and we are back to square one.
            Err((rb_eRuntimeError, cstr!("some error!")));
        }

        // `droppable` will be dropped here
    };

    match result {
        Ok(value) => return value,
        Err((exc, message)) => rb_raise(exc, message),
    }
}
```

Even then, this code still implicitly relies on several assumptions about the standard library and compiler that could break in future releases or on more exotic platforms.

This function is considered experimental. Its signature and attributes may change in future releases, or may be removed altogether. It is **highly** recommended that you never call this directly from Rust. Instead, return normally from your Rust code and use a Ruby (or C) wrapper to raise the exception for you.

Reply:
Would you be up for merging this if we add this big disclaimer?

Reply:
If we don't merge it, we'll have to remove it from Helix. Not being able to raise directly feels pretty limiting, but I agree that there are a lot of risks to it.

Reply:
Sorry I should be more clear: I was suggesting to put this into the `Safety` section. But I need to think about it more before merging (if this should be behind a feature, etc)

Reply:
I think Helix is actually in a pretty good position to do the wrapper thing, since we already generate all the code for you. Need to talk to @sgrif about his usages. (TBH, I'm not sure if all the usages there are safe either.)